docs/src/design/proposals/Registration.md

```diff
@@ -0,0 +1,518 @@
+# Introduction
+
+This document proposes a Registration API which allows for components
+of Open MCT to communicate with one another using reusable, internally
+standardized patterns.
+
+# Developer Use Cases
+
+1. Extending and maintaining Open MCT itself.
+2. Adapting and customizing Open MCT for use in specific missions.
+3. Developing features for use with Open MCT across multiple different
+   missions.
+
+# Scope
+
+As demonstrated by the existing APIs, writing plugins is sufficient to
+satisfy the three use cases above in the majority of cases. The only feature
+which is known to be unsatisfiable by plugins is plugin support itself.
+
+As such, prefer to keep "plugin-external" components small and simple, to
+keep the majority of development in plugins (or plugin-like components.)
+
+The "registration API" described in this document is limited to that scope:
+It describes classes and patterns that can allow plugins to interact,
+while making minimal assumptions about what specific functionality is to
+be implemented in these plugins.
+
+By analogy to current API, this set of functionality will effectively
+replace the [Framework Layer](http://nasa.github.io/openmctweb/guide/#framework)
+of Open MCT.
+
+# Problems to Address
+
+1. Dependencies between plugins are implicit; a plugin may fail if its
+   required dependencies are not included, without any clear indication
+   of why this failure has occurred. Significant familiarity with
+   Open MCT is typically required to debug in these circumstances.
+2. Extension points are often implicit; no specific plugin is
+   identifiably responsible for defining any specific extension category
+   or composite service. These are instead paired by string matching.
+   This makes it difficult to follow how the application is initialized
+   and how objects are passed around at run-time, and creates an
+   additional documentation burden, as these named extension points do not
+   fit into more standard API documentation.
+3. Reuse of components between plugins is limited. Exposing base classes
+   from one plugin and reusing them from another is overly-difficult.
+
+# Principles
+
+1. The Registration API is exposed as a set of classes in one or more
+   namespaces. This supports reuse and extension using standard,
+   well-known object-oriented patterns. A composition-oriented style
+   is still supported and encouraged, but not enforced.
+2. Extension points should be expressed as objects with known interfaces.
+   (More specifically, they should be _expressible_ in this fashion;
+   it is out of scope for the Registration API to expose any specific
```

Comment:
> My concerns about the facade route (or, more generally, on breaking from the "everything's just a plugin and they collaborate" paradigm, and differentiating internal API from our plugin API):
> 
> We no longer benefit (as much) from dogfooding. Currently, we are very regularly exercising the same interfaces and mechanisms for extensibility/modularity that plugins use. We have a lot of insight into how well they work and how to improve them as a consequence.

We aren't dogfooding anything that people want or need.  We're just hampering our own ability to develop simple clean code by forcing ourselves to use a weird plugin system and expecting others to use the same complex system.  More so, we're missing the forest for the trees: our application is tightly coupled and monolithic, but we claim that it's not because "everything is a plugin".

> Open MCT becomes more monolithic. Encapsulating certain platform features/behaviors as plugins takes a bit more work, but is good for long-term flexibility and maintainability. We don't expect third-party developers to swap out one version of Edit mode for another, but we know that we sometimes do that. Why should that be harder than swapping out plugins?

Don't agree; swapping out editmode was not easy.  Our plugin infrastructure serves to obscure the relationship between code, it does not help us encapsulate functionality.  This isn't dogfooding because no one else even uses it.

> This is also a lesson learned from desktop MCT - late in the project, there were major UX changes around editing/saving that weren't well-encapsulated (IIRC this was deadline-motivated, and the intent was always to refactor later - parallels to NEM and #481.) Project fell into maintenance mode, bugs in new editing/saving model were resolved in-place. By the time we recognized that there were fundamental UX issues in the new design, the scope of a fix for these issues became inestimable, and impractical to resolve incrementally.

I'm not sure this is the sort of thing that could ever be prevented by writing more code-- in fact, I think that only serves to exacerbate it. An overengineered system will lack flexibility and have trouble adapting to change.  We need to simplify and embrace change.

Should OpenMCT fall into maintenance only mode, we will have failed.  Our success is predicated on us continuously producing and adding value so that we don't ever go into maintenance mode.  I don't see this plugin system helping us do that.

> The specific dependencies of plugins remain unclear; actually, they become less clear than they currently are. You need to read through imperative code to understand a plugin's actual dependencies.

I say this again: until we have a really good *external* reason, plugins should not have dependencies on anything that is not part of the core platform.  Put another way: the only dependency a plugin has is the core platform.  And the core platform exposes a known set of services, or at least it will, when we finish our documentation.

I have yet to see a viable example of a plugin that would depend on another plugin.

> Sticking a facade on top of the architecture means that any extension points we want to utilize need to get exposed through that facade. This means either hindering ubiquitous extensibility, or creating a big interface (or set of interfaces) that is largely redundant to the architecture we're trying to stick a facade in front of.

Agree here: we just expose the architecture and point to the extension points.  Let's not hide anything.  If that architecture results in a large namespaced object e.g. `MCT.telemetryService.requestTelemetry(domainObject, options);` or `MCT.telemetryService.registerProvider(myProvider)` or `var plot = new MCT.Objects.Plot(); MCT.persistence.persist(plot)`, well, that's fine with me.  At least it's VERY clear how to use it.  

Our build system is what allows us to separate code and ensure maintainability while still exposing a clean interface.  Developers have requested specific extension points(view types).  We have seen the value in other extension points(persistence, telemetry).  We do not need to expose a confusing plugin registration API; we need to provide simple clean interfaces for plugin developers to use.